### DISCUSSION TITLE
Resolver is not firing onChange for revalidate ?

### DISCUSSION THREAD

**Discussion by tkharuk** [react-hook-form/react-hook-form#7973] [Q&A]
Hi,

been trying to apply top-level validation via `resolver`.

And currently facing an issue that it is not triggered `onChange` 
unless I'm subscribed to a `formState` where my form initiation resides.

Is this expected?

sandbox:
https://codesandbox.io/s/react-hook-form-joiresolver-v7-forked-5u211v?file=/src/index.js

**Comment by fahadsohail482** [comment]
Hi @tkharuk You need to change the `mode` to `onChange` to trigger on every change.
![image](https://user-images.githubusercontent.com/46647496/157074146-76122aea-5439-42f2-a87f-e1fb43eaca5a.png)
here is the working [SandBox](https://codesandbox.io/s/react-hook-form-joiresolver-v7-forked-05mvj8?file=/src/index.js:995-1003)
Hoping that this solves your concern.

**Reply by tkharuk** [reply]
Hi @fahadsohail482, unfortunately, this is not a good fix for me.
Default strategy of `mode: 'onSubmit', reValidateMode: 'onChange',` generally provides a better UX.

In doc example, [resolver](https://codesandbox.io/s/react-hook-form-validationschema-v7-bt78h?file=/src/index.js) works just fine without `mode` changes.

**Comment by tkharuk** [comment]
@bluebill1049 @fahadsohail482 so, is this "expected" behaviour, not a bug?
Are there any ways for me to preserve default error mode with `resolver` and `fieldState`?

**Reply by tkharuk** [reply]
bump :(
any ideas @bluebill1049 @fahadsohail482 ?

**Comment by Moshyfawn** [comment]
Hey, @tkharuk!

If you set `mode: "onSubmit"` _(default value)_, the validation is triggered only `onSubmit`, that's the expected behavior.

[In the example](https://codesandbox.io/s/react-hook-form-validationschema-v7-bt78h?file=%2Fsrc%2Findex.js) you've provided in https://github.com/react-hook-form/react-hook-form/discussions/7973#discussioncomment-2315837, the reason the validation fires is because clicking the `button` element without it's type assigned _(`type="button"`)_ triggers the form submission event, which corresponds to the `mode: "onSubmit"`.

**Reply by tkharuk** [reply]
@Moshyfawn thanks for checking.

I'm completely at a loss now...

From what I see, the default behaviour is `mode: "onSubmit"`, and what is important, `reValidateMode: 'onChange'` !

<img width="570" alt="Screenshot 2022-03-30 at 12 29 43 PM" src="https://user-images.githubusercontent.com/1205052/160799581-2d8367bd-65dc-4e81-ae8a-6f0b3f829c2c.png">

---

Here is the example from https://react-hook-form.com/get-started#SchemaValidation

As you can see, the validation works as expected - first validation **onSubmit**, and after that **onChange**:

https://user-images.githubusercontent.com/1205052/160800356-a94d0cfb-13d2-43ee-ac13-276debafe6dd.mov


---

But in [my example](https://codesandbox.io/s/react-hook-form-joiresolver-v7-forked-5u211v?file=/src/index.js), it fires only **onSubmit**:

https://user-images.githubusercontent.com/1205052/160800646-070cd425-0a40-4844-afaa-32b466837908.mov


---

However! If I read from `formState` in component, everything starts working as expected - **onSubmit** and then **onChange**:

https://user-images.githubusercontent.com/1205052/160801384-2ce1abd9-b7b8-4c6e-bd0e-01d86759bc6d.mov


---

Is this really expected behaviour?

**Reply by bluebill1049** [reply]
hey @tkharuk i just tried your example and its working fine: https://codesandbox.io/s/react-hook-form-joiresolver-v7-forked-k7egtp?file=/src/index.js

1. added the mode 'onChange'
2. change the version to use the latest instead of rc version of v7

**Reply by bluebill1049** [reply]
mode: is behavior before submit
revalidateMode: is after form is submitted

hopefully that clears things up a bit 👍

**Reply by Moshyfawn** [reply]
@tkharuk, you're using **beta** version of resolver and RHF packages. Try upgrading to a **stable** version instead
```json
"dependencies": {
  "@hookform/resolvers": "2.0.0-beta.13",
  "react-hook-form": "7.0.0-beta.11",
}
```

**Reply by tkharuk** [reply]
> mode: is behavior before submit revalidateMode: is after form is submitted
> 
> hopefully that clears things up a bit 👍

I understand this, and as I've said here before, default strategy of `mode: 'onSubmit', reValidateMode: 'onChange'` is really something I would like to have. I do not want `mode: 'onChange'`.


But, seems like version helped!
I haven't even noticed it is rc.

Ok, checking in my more complex environment (though it uses stable versions :( )!

**Comment by tkharuk** [comment]
Ok, seems like while trying to build a simpler reproduction I've introduced a similarly looking issue, and went down the wrong rabbit hole.

### But in actuality, the issue I've started with is:

- 2 fields, validated one against another
- error is shown and cleared for the field in change, not for another one

https://user-images.githubusercontent.com/1205052/160865453-a65e833b-dcaa-44b8-930c-9ac1f2528abc.mov

### Reproduction: https://codesandbox.io/s/react-hook-form-joiresolver-v7-forked-06f1uw


Again, is this something I'm using incorrectly or expected?

**Reply by Moshyfawn** [reply]
> Ok, seems like while trying to build a simpler reproduction I've introduced a similarly looking issue, and went down the wrong rabbit hole.
> ### But in actuality, the issue I've started with is:
> 
>     * 2 fields, validated one against another
> 
>     * error is shown and cleared for the field in change, not for another one
> 
> 
> resolver-errors-not-eq-formstate-errors.mov
> ### Reproduction: https://codesandbox.io/s/react-hook-form-joiresolver-v7-forked-06f1uw
> 
> Again, is this something I'm using incorrectly or expected?

To create a validation dependency between the fields, consider using [the `deps` prop](https://react-hook-form.com/api/useform/register) and pass it an array of field names you need to be validated along with the input you're defining the `deps` array in.  
```jsx
<input
  {...register("test", {
    deps: ['inputA', 'inputB'],
  })}
/>
```

**Reply by tkharuk** [reply]
Tanks, this helped!

I thought you can **not** use `deps` with `resolver`:
<img width="926" alt="Screenshot 2022-03-30 at 6 50 21 PM" src="https://user-images.githubusercontent.com/1205052/160877422-3bd4f61c-487c-4375-b4a2-a026885c26cd.png">


#### From these rules I thought that:
- revalidation should come from `errors` object
- you can not use `deps` because for me it is the same cohort as `required` since it is a part of `rules` property
  - `useController({ name, rules: { required: true, deps: 'a' } });`

False assumptions, I suppose :(
Also, in my case, `deps` would need to be generated based on `useFieldArray` length, making it a bit weird...

**Reply by tkharuk** [reply]
@[Moshyfawn](https://github.com/Moshyfawn) yeah, mixing up top-level validation with inline-level `deps` looks very weird, two sources of truth when it comes to validations. 
Also in my case inline validations does not play well with reusability - in some scenarios fields are dependent and in some - not. 

I just find that **top-level validations** give much more flexibility and control.

Any hints on how could I achieve desired outcome with resolver or maybe something else?

I'm trying to evaluate `react-hook-form` for our needs, and testing it against a mid-complex form:
<img width="664" alt="Screenshot 2022-03-31 at 1 44 21 PM" src="https://user-images.githubusercontent.com/1205052/161037689-d9ee163f-7baf-4dcd-b660-a4910364ca3d.png">

- 3 level of array fields nesting
- validations on field level
- validations between fields of same level
- validations between fields of different levels

And whatever I try with `react-hook-library` just doesn't "click".
Maybe I'm just doing it fully wrong, I don't know.

I really want to "make it work" though.
Next step for me would be to fork library and try to play my examples and debug a bit.
